### PR TITLE
Take selinuxd into use

### PR DESCRIPTION
**Pull request by jhrozek**
Changes SELinux integration by ditching the intermediate configMaps 
previously used by the CM reconcile loop of the selinux controller. 
Instead, the daemon running in the DS now takes an boolean argument that 
signifies if the SELinux feature should be enabled. This is false by 
default.

If true, the DS deploys an additional container running selinuxd which 
listens to any files under /etc/selinux.d and can be queried for status 
over a UNIX socket mounted under /var/run/selinuxd. Both the socket 
directory and the drop-in directory are EmptyDirs shared between selinuxd
and the controller.

In addition, the operator receives a --with-selinux=true flag, which makes
the operator create an additional manager that listens for the 
selinuxpolicy API object. Contents of those objects are reconciled to the
/etc/selinux.d drop-in directory and then the controller queries the state
of selinuxd to make sure the policy was indeed added or removed.

### Release note
```release-note
The  selinux component can now be enabled or disabled through the  CongfiMap named  config by toggling a boolean option called EnableSelinux
 
Since not all Linux distributions support SeLinux, its support is disabled by default.
```

TODO:
   - [ ] The SELinux controller should be on by default for OpenShift
         through kustomize
   - [ ] optimize writing the policy files to write only if the file
          content differs when selinux'd `/ready` endpoint merges
   - [ ] take errors from selinuxd into use, extend the status to present them and add a test to exercise a failing selinux policy
   - [ ] the selinuxd container currently sets no resource requests/limits
   - [ ] make the selinuxd container read-only, needs to mount the selinuxd database in a separate volume
   - [ ] open a ticket to consider adding events

/kind feature

### CODE REVIEW

**Comment by JAORMX** [comment]
@jhrozek can you add a release note about the flag to enable/disable SELinux?

File: deploy/base/kustomization.yaml
Comment (JAORMX) [inline deploy/base/kustomization.yaml:33]: let's unify this by adding the `udica` dependency to the `selinuxd` image. This way we don't need the extra image.

File: deploy/base/operator.yaml
Comment (JAORMX) [inline deploy/base/operator.yaml:37]: same comment as above.

File: cmd/security-profiles-operator/main.go
Comment (JAORMX) [inline cmd/security-profiles-operator/main.go:44]: This flag doesn't seem to be used anywhere.

File: deploy/base/config.yaml
Comment (JAORMX) [inline deploy/base/config.yaml:10]: Let's add `EnableSelinux: "false"` here too. It's better to document all the available options.

**Review by JAORMX — CHANGES_REQUESTED** [review]

File: test/e2e_test.go
Comment (JAORMX) [inline test/e2e_test.go:149]: This `sed` could be simplified once you always add `EnableSelinux` in the sample config.

File: deploy/base/kustomization.yaml
Comment (jhrozek) [reply deploy/base/kustomization.yaml:33]: I'm not sure I understand what you are proposing. Copy the files from `udica` to the `selinuxd` image as part of the image build process and then use `selinuxd` as the init container that primes the selinux drop-directory?

File: cmd/security-profiles-operator/main.go
Comment (jhrozek) [reply cmd/security-profiles-operator/main.go:44]: Yeah, I think I'll just remove it from now. In fact, I think I might even remove the EnableSeccomp flag and make only SELinux optional until we figure out a nicer way to instantiate the DS based on options.

File: deploy/base/config.yaml
Comment (jhrozek) [reply deploy/base/config.yaml:10]: Will do.

File: test/e2e_test.go
Comment (jhrozek) [reply test/e2e_test.go:149]: Will do.

File: deploy/base/kustomization.yaml
Comment (JAORMX) [reply deploy/base/kustomization.yaml:33]: in the selinuxd Dockerfile:
```
microdnf install -y udica
```
That would basically install the base policies that `udica` comes with.
Then we could use that same image as the init container that primes the drop-directory.

Comment (jhrozek) [reply deploy/base/kustomization.yaml:33]: oh, right. Packages still exist :-)

File: test/e2e_test.go
Comment (jhrozek) [reply test/e2e_test.go]: Done

File: deploy/base/config.yaml
Comment (jhrozek) [reply deploy/base/config.yaml]: Done

File: cmd/security-profiles-operator/main.go
Comment (jhrozek) [reply cmd/security-profiles-operator/main.go]: Removed

File: deploy/base/operator.yaml
Comment (jhrozek) [reply deploy/base/operator.yaml:35]: Removed

File: deploy/base/kustomization.yaml
Comment (jhrozek) [reply deploy/base/kustomization.yaml:31]: Done in selinuxd.
Here's a follow-up question: I think we can ditch the second init container entirely and change the ownership of `/etc/selinux.d` in the init container. Then we could copy the files into the directory when selinuxd starts with a lifecycle handler. It is maybe a bit hacky to have the logic split over two places, but at least it would allow us to keep the DS template more minimal and better manageable.

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (JAORMX) [inline internal/pkg/controllers/spod/bindata/spod.go:160]: Why was the DAC_OVERRIDE capability needed again?

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (JAORMX) [inline internal/pkg/controllers/spod/bindata/spod.go:239]: let's switch to `jaosorior/selinuxd` as that's what's currently being built automatically when we push to the selinuxd repo. We should now add a github issue about moving that to an appropriate org. (First we gotta figure what that org is...)

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (JAORMX) [inline internal/pkg/controllers/spod/bindata/spod.go:271]: Let's open up a Github issue for this. We could either make that directory configurable, or just expect a directory inside of `/var/run`.

File: deploy/base/config.yaml
Comment (saschagrunert) [inline deploy/base/config.yaml:10]: Can we directly use a bool here?
```suggestion
  EnableSelinux: false
```

**Review by saschagrunert — APPROVED** [review]

File: deploy/base/config.yaml
Comment (jhrozek) [reply deploy/base/config.yaml:10]: Haha, yes, I'll see about fixing this

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (jhrozek) [reply internal/pkg/controllers/spod/bindata/spod.go:239]: Oops, sorry. This was a leftover from when I was testing some custom image of my own.

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (jhrozek) [reply internal/pkg/controllers/spod/bindata/spod.go:160]: I..might have copied the capabilities of the other init container blindly. I'll test what is really needed. But yeah, DAC_OVERRIDE might not be since we're not accessing anything the usual UNIX permissions wouldn't allow us to access.

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:229]: Yes, I planned to iterate on this PR by taking the newer features of selinuxd into account in subsequent PRs.

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:295]: I don't think that feature was available in selinuxd at that point. But it seems easy enough to add, I'll see about adding it to this PR.

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:209]: It does trigger an inotify. I'll add a check if the contents are already written before writing the file (I'll add a simple check for now, just comparing the contents)

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (jhrozek) [reply internal/pkg/controllers/spod/bindata/spod.go:271]: I'll do that

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (naveensrinivasan) [inline internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:169]: The reconciliation isn't reporting any k8s events. k8s events help with troubleshooting for the operators compared to logs. It also helps e2e tests to validate the condition.

Could we add events for the invalid and valid conditions?

Comment (JAORMX) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:169]: This PR is already quite big. Mind if we file a Github issue and do that in a subsequent PR?

**Comment by saschagrunert** [comment]
Unfortunately this now needs a rebase after the merge of https://github.com/kubernetes-sigs/security-profiles-operator/pull/179

File: deploy/base/kustomization.yaml
Comment (jhrozek) [reply deploy/base/kustomization.yaml:31]: As discussed on Slack, let's do this if/when we need to make the DS manageable better. I added a comment in the meantime.

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:231]: So I tried removing the write on every reconcile but found a race in selinuxd instead. There is a PR for selinuxd to address it now, but so far I submit this PR with the write on every reconcile and would take the selinuxd fixed into effect once that PR is merged.

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go]: Done

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go]: Done, but we still need to extend the status subresource and add a test. So far I added a TODO item into the PR comment and will file issues for the follow-up work.

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (jhrozek) [reply internal/pkg/controllers/spod/bindata/spod.go]: Dropped.

File: internal/pkg/controllers/spod/bindata/spod.go
Comment (jhrozek) [reply internal/pkg/controllers/spod/bindata/spod.go]: Fixed.

File: deploy/base/config.yaml
Comment (jhrozek) [reply deploy/base/config.yaml:10]: Hmm, I must be misunderstanding something. I'm getting errors trying to use a straight-up boolean in a configmap, seems like CMs only allow string values?

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (jhrozek) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:184]: I added a todo note to the opening comment and I'll open all the tickets at the same time

File: test/suite_test.go
Comment (JAORMX) [inline test/suite_test.go:166]: you need a trailing `,`

**Review by JAORMX — CHANGES_REQUESTED** [review]

File: test/suite_test.go
Comment (jhrozek) [reply test/suite_test.go]: I like to live dangerously and submit patches to CI without even running make.

**Review by JAORMX — APPROVED** [review]
/approve

will wait for @saschagrunert to add lgtm

File: internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go
Comment (naveensrinivasan) [reply internal/pkg/controllers/selinuxpolicy/selinuxpolicy_controller.go:185]: That works me. Thanks 👍

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**
